doc/configuration.md
# Configuration
All configurations for the service can be set using command line flags and can be displayed via `--help`. The configuration options will be described for each component individually.

## TODO/Questions
- [ ] I think we should only add the Discovery specific flags. The "problem" is that several flags are added for each component, e.g., db-X, but are not needed for all components. Are the only necessary flags the ones which can be found [here](../server/commands/discovery/discovery.go)
- [ ] Sollten nicht alle Service Optionen auch einen flag für den service haben? ZB WithoutEvidenceStore hat keinen.


## Discovery

```
Usage:
  discovery [flags]

Flags:
      --api-cors-allowed-headers stringArray   Specifies the headers allowed in CORS (default [Content-Type,Accept,Authorization])
      --api-cors-allowed-methods stringArray   Specifies the methods allowed in CORS (default [GET,POST,PUT,DELETE])
      --api-cors-allowed-origins stringArray   Specifies the origins allowed in CORS
      --api-default-password string            Specifies the default API password (default "clouditor")
      --api-default-user string                Specifies the default API username (default "clouditor")
      --api-grpc-port uint16                   Specifies the port used for the Clouditor gRPC API (default 9091)
      --api-http-port uint16                   Specifies the port used for the Clouditor HTTP API (default 8081)
      --api-jwks-url string                    Specifies the JWKS URL used to verify authentication tokens in the gRPC and HTTP API (default "http://localhost:8080/v1/auth/certs")
      --api-key-password string                Specifies the password used to protect the API private key (default "changeme")
      --api-key-path string                    Specifies the location of the API private key (default "~/.clouditor/api.key")
      --api-key-save-on-create                 Specifies whether the API key should be saved on creation. It will only created if the default location is used. (default true)
      --assessment-url string                  Specifies the Assessment URL (default "localhost:9093")
      --certification-target-id string         Specifies the Certification Target ID (default "00000000-0000-0000-0000-000000000000")
      --dashboard-callback-url string          The callback URL of the Clouditor Dashboard. If the embedded server is used, a public OAuth 2.0 client based on this URL will be added (default "http://localhost:8080/callback")
      --db-host string                         Provides address of database (default "localhost")
      --db-in-memory                           Uses an in-memory database which is not persisted at all
      --db-name string                         Provides name of database (default "postgres")
      --db-password string                     Provides password of database (default "postgres")
      --db-port uint16                         Provides port for database (default 5432)
      --db-ssl-mode string                     The SSL mode for the database (default "disable")
      --db-user-name string                    Provides user name of database (default "postgres")
      --discovery-auto-start                   Automatically start the discovery when engine starts
      --discovery-csaf-domain string           The domain to look for a CSAF provider, if the CSAF discovery is enabled
  -p, --discovery-provider strings             Providers to discover, separated by comma
      --discovery-resource-group string        Limit the scope of the discovery to a resource group (currently only used in the Azure discoverer
  -h, --help                                   help for discovery
      --log-level string                       The default log level (default "info")
      --service-oauth2-client-id string        Specifies the OAuth 2.0 client ID (default "clouditor")
      --service-oauth2-client-secret string    Specifies the OAuth 2.0 client secret (default "clouditor")
      --service-oauth2-token-endpoint string   Specifies the OAuth 2.0 token endpoint (default "http://localhost:8080/v1/auth/token")
```

internal/config/config.go
// Copyright 2024 Fraunhofer AISEC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
//           $$\                           $$\ $$\   $$\
//           $$ |                          $$ |\__|  $$ |
//  $$$$$$$\ $$ | $$$$$$\  $$\   $$\  $$$$$$$ |$$\ $$$$$$\    $$$$$$\   $$$$$$\
// $$  _____|$$ |$$  __$$\ $$ |  $$ |$$  __$$ |$$ |\_$$  _|  $$  __$$\ $$  __$$\
// $$ /      $$ |$$ /  $$ |$$ |  $$ |$$ /  $$ |$$ |  $$ |    $$ /  $$ |$$ | \__|
// $$ |      $$ |$$ |  $$ |$$ |  $$ |$$ |  $$ |$$ |  $$ |$$\ $$ |  $$ |$$ |
// \$$$$$$\  $$ |\$$$$$   |\$$$$$   |\$$$$$$  |$$ |  \$$$   |\$$$$$   |$$ |
//  \_______|\__| \______/  \______/  \_______|\__|   \____/  \______/ \__|
//
// This file is part of Clouditor Community Edition.

package config

import (
	"strings"

	"github.com/spf13/cobra"
	"github.com/spf13/viper"
	"golang.org/x/oauth2/clientcredentials"
)

const (
	APIDefaultUserFlag                   = "api-default-user"
	APIDefaultPasswordFlag               = "api-default-password"
	APIKeyPasswordFlag                   = "api-key-password"
	APIKeyPathFlag                       = "api-key-path"
	APIKeySaveOnCreateFlag               = "api-key-save-on-create"
	APIgRPCPortFlag                      = "api-grpc-port"
	APIHTTPPortFlag                      = "api-http-port"
	APICORSAllowedOriginsFlags           = "api-cors-allowed-origins"
	APICORSAllowedHeadersFlags           = "api-cors-allowed-headers"
	APICORSAllowedMethodsFlags           = "api-cors-allowed-methods"
	APIJWKSURLFlag                       = "api-jwks-url"
	EmbeddedOAuth2ServerEnabledFlag      = "embedded-oauth2-server-enabled"
	EmbeddedOAuth2ServerPublicURLFlag    = "embedded-oauth2-server-public-url"
	ServiceOAuth2EndpointFlag            = "service-oauth2-token-endpoint"
	ServiceOAuth2ClientIDFlag            = "service-oauth2-client-id"
	ServiceOAuth2ClientSecretFlag        = "service-oauth2-client-secret"
	CertificationTargetIDFlag            = "certification-target-id"
	AssessmentURLFlag                    = "assessment-url"
	OrchestratorURLFlag                  = "orchestrator-url"
	EvidenceStoreURLFlag                 = "evidence-store-url"
	DBUserNameFlag                       = "db-user-name"
	DBPasswordFlag                       = "db-password"
	DBHostFlag                           = "db-host"
	DBNameFlag                           = "db-name"
	DBPortFlag                           = "db-port"
	DBSSLModeFlag                        = "db-ssl-mode"
	DBInMemoryFlag                       = "db-in-memory"
	CreateDefaultCertificationTargetFlag = "create-default-certification-target"
	DiscoveryAutoStartFlag               = "discovery-auto-start"
	DiscoveryProviderFlag                = "discovery-provider"
	DiscoveryResourceGroupFlag           = "discovery-resource-group"
	DiscoveryCSAFDomainFlag              = "discovery-csaf-domain"
	DashboardCallbackURLFlag             = "dashboard-callback-url"
	LogLevelFlag                         = "log-level"

	DefaultAPIDefaultUser                     = "clouditor"
	DefaultAPIDefaultPassword                 = "clouditor"
	DefaultAPIgRPCPort                 uint16 = 9090
	DefaultAPIgRPCPortOrchestrator     uint16 = 9090
	DefaultAPIgRPCPortDiscovery        uint16 = 9091
	DefaultAPIgRPCPortEvidenceStore    uint16 = 9092
	DefaultAPIgRPCPortAssessment       uint16 = 9093
	DefaultAPIgRPCPortEvaluation       uint16 = 9094
	DefaultAPIHTTPPortOrchestrator     uint16 = 8080
	DefaultAPIHTTPPortDiscovery        uint16 = 8081
	DefaultAPIHTTPPortEvidenceStore    uint16 = 8082
	DefaultAPIHTTPPortAssessment       uint16 = 8083
	DefaultAPIHTTPPortEvaluation       uint16 = 8084
	DefaultEmbeddedOAuth2ServerEnabled        = true
	DefaultServiceOAuth2Endpoint              = "http://localhost:8080/v1/auth/token"
	DefaultServiceOAuth2ClientID              = "clouditor"
	DefaultServiceOAuth2ClientSecret          = "clouditor"
	DefaultOrchestratorURL                    = "localhost:9090"
	DefaultEvidenceStoreURL                   = "localhost:9090"
	DefaultAssessmentURL                      = "localhost:9090"
	DefaultDBUserName                         = "postgres"
	DefaultDBPassword                         = "postgres"
	DefaultDBHost                             = "localhost"
	DefaultDBName                             = "postgres"
	DefaultDBPort                      uint16 = 5432
	DefaultDBSSLMode                          = "disable"
	DefaultDBInMemory                         = false
	DefaultCreateDefaultTarget                = true
	DefaultDiscoveryAutoStart                 = false
	DefaultDiscoveryResourceGroup             = ""
	DefaultCSAFDomain                         = ""
	DefaultDashboardCallbackURL               = "http://localhost:8080/callback"
	DefaultLogLevel                           = "info"

	EnvPrefix = "CLOUDITOR"
)

var (
	// DefaultAllowedOrigins contains a nil slice, as per default, no origins are allowed.
	DefaultAllowedOrigins []string = nil

	// DefaultAllowedHeaders contains sensible defaults for the Access-Control-Allow-Headers header.
	// Please adjust accordingly in production using WithAllowedHeaders.
	DefaultAllowedHeaders = []string{"Content-Type", "Accept", "Authorization"}

	// DefaultAllowedMethods contains sensible defaults for the Access-Control-Allow-Methods header.
	// Please adjust accordingly in production using WithAllowedMethods.
	DefaultAllowedMethods = []string{"GET", "POST", "PUT", "DELETE"}

	// DefaultAPIHTTPPort specifies the default port for the REST API.
	DefaultAPIHTTPPort uint16 = 8080
)

const (
	// DefaultCertificationTargetID is the default certification target ID. Currently, our discoverers have no way to differentiate between different
	// targets, but we need this feature in the future. This serves as a default to already prepare the necessary
	// structures for this feature.
	DefaultCertificationTargetID = "00000000-0000-0000-0000-000000000000"

	// DefaultEvidenceCollectorToolID is the default evidence collector tool ID.
	DefaultEvidenceCollectorToolID = "Clouditor Evidences Collection"
)

func init() {
	cobra.OnInitialize(InitConfig)
}

func InitCobra(engineCmd *cobra.Command) {
	engineCmd.Flags().Bool(EmbeddedOAuth2ServerEnabledFlag, DefaultEmbeddedOAuth2ServerEnabled, "Specifies whether the embedded OAuth 2.0 authorization server is started as part of the REST gateway. For production workloads, an external authorization server is recommended.")

	_ = viper.BindPFlag(EmbeddedOAuth2ServerEnabledFlag, engineCmd.Flags().Lookup(EmbeddedOAuth2ServerEnabledFlag))
}

func InitConfig() {
	viper.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
	viper.SetEnvPrefix(EnvPrefix)
	viper.SetConfigName("clouditor")
	viper.SetConfigType("yaml")
	viper.AddConfigPath(".")
	viper.AutomaticEnv()

	_ = viper.ReadInConfig()
}

// ClientCredentials configures the OAuth 2.0 client credentials for a service
func ClientCredentials() *clientcredentials.Config {
	return &clientcredentials.Config{
		ClientID:     viper.GetString(ServiceOAuth2ClientIDFlag),
		ClientSecret: viper.GetString(ServiceOAuth2ClientSecretFlag),
		TokenURL:     viper.GetString(ServiceOAuth2EndpointFlag),
	}
}
